doc/ubertooth-btle.1
.\" generated with Ronn/v0.7.3
.\" http://github.com/rtomayko/ronn/tree/0.7.3
.
.TH "UBERTOOTH\-BTLE" "1" "March 2017" "" ""
.
.SH "NAME"
\fBubertooth\-btle\fR \- Bluetooth Low Energy (BLE) sniffing and more
.
.SH "SYNOPSIS"
.
.nf

ubertooth\-btle \-f [\-A 37|38|39] [\-r output\.pcapng]
.
.fi
.
.SH "DESCRIPTION"
\fBubertooth\-btle\fR is a tool for doing Fun Stuff(TM) with BLE\. It can do the following things:
.
.IP "\(bu" 4
Sniff connections
.
.IP "\(bu" 4
Interfere with connections
.
.IP "\(bu" 4
Send advertising packets (experimental)
.
.IP "" 0
.
.P
Sniffing connections is the most robust feature supported by \fBubertooth\-btle\fR\. It has two primary modes of operation: follow mode and promiscuous mode\.
.
.P
Follow mode is the preferred mode for general use\. In this mode, Ubertooth will listen on one of three advertising channels waiting for a BLE connection to be established\. When a connection is established, Ubertooth will hop along the data channels, passively capturing the data sent between the central and peripheral\. After the connection terminates, Ubertooth will return to the advertising channel and wait for another connection\.
.
.P
Promiscuous mode is an experimental mode for sniffing connections after they have already been established\. This mode can be used to sniff long\-lived connections\.
.
.P
When sniffing, Ubertooth can only operate in either follow mode or promiscuous mode, but not both at the same time\. If you are unsure which mode to use, use follow mode\.
.
.P
By default, Ubertooth will follow any connection it observes\. You can limit this to following a specific Bluetooth Address (BD ADDR) using the \fB\-t\fR command line flag\. For example, the following command will only sniff connections where the central or peripheral\'s BD ADDR is \fB22:44:66:88:AA:CC\fR:
.
.IP "" 4
.
.nf

ubertooth\-btle \-f \-t22:44:66:88:AA:CC
.
.fi
.
.IP "" 0
.
.P
In both sniffing modes, Ubertooth can log data to PCAP or PcapNG with a variety of pseudoheaders\. The recommended logging format is PcapNG (\fB\-r\fR) or PCAP with LE Pseudoheader (\fB\-q\fR)\. For compatibility with crackle (see \fIUSING WITH CRACKLE\fR), use PCAP with PPI (\fB\-c\fR)\.
.
.P
Interfering with connections is a feature for causing intentional interference with newly established or long\-lived connections\. When this attack succeeds, the BLE connection between the central and peripheral will be terminated\. Pair the \fB\-i\fR or \fB\-I\fR flag with \fB\-f\fR to interfere with new connections or \fB\-p\fR to interfere with long\-lived connections\. Note that causing intentional interference may be illegal in your jurisdiction\. Check your local laws before using this feature\.
.
.P
Finally, \fBubertooth\-btle\fR supports transmitting advertising packets with a specified BD ADDR\. This feature, referred to as faux slave mode, is experimental and may not function as intended\. Use at your own risk\.
.
.SH "EXAMPLES"
Sniff all connections on advertising channel 38, logging all data to PcapNG:
.
.IP "" 4
.
.nf

ubertooth\-btle \-f \-A 38 \-r log\.pcapng
.
.fi
.
.IP "" 0
.
.P
Interfere with connections recovered with promiscuous mode:
.
.IP "" 4
.
.nf

ubertooth\-btle \-p \-I
.
.fi
.
.IP "" 0
.
.P
Send advertising packets using BD ADDR \fB22:44:66:88:AA:CC\fR:
.
.IP "" 4
.
.nf

ubertooth\-btle \-s22:44:66:88:AA:CC
.
.fi
.
.IP "" 0
.
.SH "OPTIONS"
.
.TP
\fB\-h\fR
Displays help message
.
.P
Major modes:
.
.TP
\fB\-f\fR
Follow mode: sniff connections as they are established
.
.TP
\fB\-p\fR
Promiscuous mode: sniff already\-established connections
.
.TP
\fB\-s<BD ADDR>\fR
Inject advertising packets using specified BD ADDR
.
.P
Interference (pair with \fB\-f\fR or \fB\-p\fR):
.
.TP
\fB\-i\fR
Interfere with one connection and return to idle
.
.TP
\fB\-I\fR
Interfere continuously with many connections
.
.P
Filtering:
.
.TP
\fB\-t<BD ADDR>\fR
Limit connection following and interference in follow mode to the specified BD ADDR
.
.P
Logging:
.
.TP
\fB\-r <output\.pcapng>\fR
Log to PcapNG (preferred)
.
.TP
\fB\-q <output\.pcap>\fR
Log to PCAP with \fBDLT_BLUETOOTH_LE_LL_WITH_PHDR\fR
.
.TP
\fB\-c <output\.pcap>\fR
Log to PCAP with PPI (for compatibility with crackle(1))
.
.P
Miscellaneous:
.
.TP
\fB\-A <37|38|39>\fR
Which advertising channel to use in follow mode (default: 37)
.
.TP
\fB\-a[address]\fR
Get or set access address in promiscuous mode
.
.TP
\fB\-v[01]\fR
Get or set CRC verification (default: 0)
.
.TP
\fB\-x<0\-32>\fR
Allow n access address violations (default: 32)\. Filtering occurs on host\.
.
.P
Data source:
.
.TP
\fB\-U<0\-7>\fR
Which Ubertooth to use
.
.SH "USING WITH CRACKLE"
\fBcrackle\fR is a tool for cracking the BLE key exchange and decrypting encrypted data\. To capture data for use with \fBcrackle\fR, sniff connections in follow mode using \fB\-f\fR and log data to PCAP/PPI using \fB\-c\fR\. Example:
.
.IP "" 4
.
.nf

ubertooth\-btle \-f \-c crack\.pcap
.
.fi
.
.IP "" 0
.
.P
Refer to \fBcrackle\fR documentation for further details\.
.
.SH "SEE ALSO"
crackle(1): https://github\.com/mikeryan/crackle
.
.SH "COPYRIGHT"
\fBubertooth\-btle\fR is Copyright (C) 2012\-2017 Mike Ryan\. This tool is released under the GPLv2\. Refer to COPYING for futher details\.


changelog
ubertooth (2017-03-R2) unstable; urgency=low
  * USB API versioning - We have added USB API versioning to improve the
    experience of failures due to mismatched firmware and host code. You may
    find that Ubertooth tools complain that your firmware is out of date.
    Updating firmware is simple and reliable, please follow the instructions
    on the wiki. A release archive with precompiled firmware is available.
  * Basic Rate PCAP output - the pcap file produced by the Ubertooth tools
    should all be read by the latest Wireshark versions without the need to
    install plugins.
  * Kismet plugins - Kismet is currently under heavy development and it has
    not been possible to continue tracking the master branch. The legacy
    Kismet plugins should continue to work, but we will wait to track git
    master until development has stabalised.
  
 -- Dominic Spill <dominicgs@gmail.com>  01 Mar 2017 10:39:00 -0700   
 
 ubertooth (2015-10-R1) unstable; urgency=low
  * This is a minor bugfix and update release
  * Update ubertooth-scan to provide more consistent output
  * It is now possible to build receive only firmware
  * Host code versioning is more reliable
  * Python tools now support Python 3. Thanks Hannes!

  
 -- Dominic Spill <dominicgs@gmail.com>  12 Oct 2015 03:10:00 -0700   
 
 ubertooth (2015-09-R2) unstable; urgency=low
  * Basic Rate Bluetooth Improvements
      * Stability has improved in performance constrained environments
      * The "uh oh, buffer not emptied" issue no longer causes the tools to loop
        indefinitely
      * AFH support has greatly improved, UAP and CLK values are now found more
        quickly when AFH is used
  * Basic Rate PCAP support
      * ubertooth-rx and ubertooth-follow support a new PCAP file format
      * The new PCAP format is also supported with a new Wireshark plugin: btbredr
      * The PCAP format is still in the early stages, but we hope to have the plugin
        included in Wireshark later this year
      * The older btbb format is still produced by the Kismet plugins, they will
        be migrated over to the new format in the next release
  * Bluetooth Smart (BLE) Improvements
      * Improved packet reception
      * Promiscuous sniffing mode is more reliable
      * Promiscuous sniffing reports the recovered parameters to the host
  * Skateboard Hacking
      * Support for interfering with connections of BLE enabled skateboards
      * ubertooth-ego targets Yuneec E-go (not Bluetooth)
      * Always hack safely and responsibly!

  
 -- Dominic Spill <dominicgs@gmail.com>  05 Sep 2015 05:30:00 -0700   

ubertooth (2014-02-R2) unstable; urgency=low
  * Bluetooth Smart (Low Energy) Support
      * Promiscuous and follow modes
      * Pcap format packet logging
      * Pairing / encryption support when paired with crackle
      * Credit for BLE features goes to Mike Ryan
  * Unified host tool for monitoring Basic Rate
      * ubertooth-rx replaces -lap, -uap, -hop tools
      * Once UAP is discovered, ubertooth-rx automatically tries to find clock
        values and begin hopping
      * Thanks to Will Code for working on this
  * Survey tool - ubertooth-scan
      * Combining both Ubertooth and a standard Bluetooth dongle
      * Ubertooth scans for non-discoverable master devices
      * Dongle probes devices for piconet information and features
  * Cmake now used for the build system
      * Improves support for non-Linux operating systems
      * More sensible handling of dependencies
      * New build instructions
  * Packaging (Experimental)
      * Early stage support for packaging systems
      * libbtbb in Homebrew repository, Ubertooth coming soon
      * MacPorts availability is under test
      * Release already available in Pentoo
  * GitHub migration
      * libbtbb, Ubertooth and gr-bluetooth all hosted on GitHub
      * Allows for more open development and collaboration model
      * Already seeing an increase in issue reporting and pull requests


 -- Dominic Spill <dominicgs@gmail.com>  20 Feb 2014 06:06:00 -0700   


host/ubertooth-tools/src/ubertooth-btle.c
/*
 * Copyright 2012-2016 Michael Ryan
 *
 * This file is part of Project Ubertooth.
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2, or (at your option)
 * any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; see the file COPYING.  If not, write to
 * the Free Software Foundation, Inc., 51 Franklin Street,
 * Boston, MA 02110-1301, USA.
 */

#include "ubertooth.h"
#include "ubertooth_callback.h"
#include <ctype.h>
#include <err.h>
#include <getopt.h>
#include <string.h>
#include <unistd.h>
#include <stdlib.h>

int convert_mac_address(char *s, uint8_t *o) {
	int i;

	// validate length
	if (strlen(s) != 6 * 2 + 5) {
		printf("Error: MAC address is wrong length\n");
		return 0;
	}

	// validate hex chars and : separators
	for (i = 0; i < 6*3; i += 3) {
		if (!isxdigit(s[i]) ||
			!isxdigit(s[i+1])) {
			printf("Error: MAC address contains invalid character(s)\n");
			return 0;
		}
		if (i < 5*3 && s[i+2] != ':') {
			printf("Error: MAC address contains invalid character(s)\n");
			return 0;
		}
	}

	// sanity: checked; convert
	for (i = 0; i < 6; ++i) {
		unsigned byte;
		sscanf(&s[i*3], "%02x",&byte);
		o[i] = byte;
	}

	return 1;
}

static void usage(void)
{
	printf("ubertooth-btle - passive Bluetooth Low Energy monitoring\n");
	printf("Usage:\n");
	printf("\t-h this help\n");
	printf("\n");
	printf("    Major modes:\n");
	printf("\t-f follow connections\n");
	printf("\t-p promiscuous: sniff active connections\n");
	printf("\t-a[address] get/set access address (example: -a8e89bed6)\n");
	printf("\t-s<address> faux slave mode, using MAC addr (example: -s22:44:66:88:aa:cc)\n");
	printf("\t-t<address> set connection following target (example: -t22:44:66:88:aa:cc)\n");
	printf("\n");
	printf("    Interference (use with -f or -p):\n");
	printf("\t-i interfere with one connection and return to idle\n");
	printf("\t-I interfere continuously\n");
	printf("\n");
	printf("    Data source:\n");
	printf("\t-U<0-7> set ubertooth device to use\n");
	printf("\n");
	printf("    Misc:\n");
	printf("\t-r<filename> capture packets to PCAPNG file\n");
	printf("\t-q<filename> capture packets to PCAP file (DLT_BLUETOOTH_LE_LL_WITH_PHDR)\n");
	printf("\t-c<filename> capture packets to PCAP file (DLT_PPI)\n");
	printf("\t-A<index> advertising channel index (default 37)\n");
	printf("\t-v[01] verify CRC mode, get status or enable/disable\n");
	printf("\t-x<n> allow n access address offenses (default 32)\n");

	printf("\nIf an input file is not specified, an Ubertooth device is used for live capture.\n");
	printf("In get/set mode no capture occurs.\n");
}

int main(int argc, char *argv[])
{
	int opt;
	int do_follow, do_promisc;
	int do_get_aa, do_set_aa;
	int do_crc;
	int do_adv_index;
	int do_slave_mode;
	int do_target;
	enum jam_modes jam_mode = JAM_NONE;
	int ubertooth_device = -1;
	ubertooth_t* ut = ubertooth_init();

	btle_options cb_opts = { .allowed_access_address_errors = 32 };

	int r;
	u32 access_address;
	uint8_t mac_address[6] = { 0, };

	do_follow = do_promisc = 0;
	do_get_aa = do_set_aa = 0;
	do_crc = -1; // 0 and 1 mean set, 2 means get
	do_adv_index = 37;
	do_slave_mode = do_target = 0;

	while ((opt=getopt(argc,argv,"a::r:hfpU:v::A:s:t:x:c:q:jJiI")) != EOF) {
		switch(opt) {
		case 'a':
			if (optarg == NULL) {
				do_get_aa = 1;
			} else {
				do_set_aa = 1;
				sscanf(optarg, "%08x", &access_address);
			}
			break;
		case 'f':
			do_follow = 1;
			break;
		case 'p':
			do_promisc = 1;
			break;
		case 'U':
			ubertooth_device = atoi(optarg);
			break;
		case 'r':
			if (!ut->h_pcapng_le) {
				if (lell_pcapng_create_file(optarg, "Ubertooth", &ut->h_pcapng_le)) {
					err(1, "lell_pcapng_create_file: ");
				}
			}
			else {
				printf("Ignoring extra capture file: %s\n", optarg);
			}
			break;
		case 'q':
			if (!ut->h_pcap_le) {
				if (lell_pcap_create_file(optarg, &ut->h_pcap_le)) {
					err(1, "lell_pcap_create_file: ");
				}
			}
			else {
				printf("Ignoring extra capture file: %s\n", optarg);
			}
			break;
		case 'c':
			if (!ut->h_pcap_le) {
				if (lell_pcap_ppi_create_file(optarg, 0, &ut->h_pcap_le)) {
					err(1, "lell_pcap_ppi_create_file: ");
				}
			}
			else {
				printf("Ignoring extra capture file: %s\n", optarg);
			}
			break;
		case 'v':
			if (optarg)
				do_crc = atoi(optarg) ? 1 : 0;
			else
				do_crc = 2; // get
			break;
		case 'A':
			do_adv_index = atoi(optarg);
			if (do_adv_index < 37 || do_adv_index > 39) {
				printf("Error: advertising index must be 37, 38, or 39\n");
				usage();
				return 1;
			}
			break;
		case 's':
			do_slave_mode = 1;
			r = convert_mac_address(optarg, mac_address);
			if (!r) {
				usage();
				return 1;
			}
			break;
		case 't':
			do_target = 1;
			r = convert_mac_address(optarg, mac_address);
			if (!r) {
				usage();
				return 1;
			}
			break;
		case 'x':
			cb_opts.allowed_access_address_errors = (unsigned) atoi(optarg);
			if (cb_opts.allowed_access_address_errors > 32) {
				printf("Error: can tolerate 0-32 access address bit errors\n");
				usage();
				return 1;
			}
			break;
		case 'i':
		case 'j':
			jam_mode = JAM_ONCE;
			break;
		case 'I':
		case 'J':
			jam_mode = JAM_CONTINUOUS;
			break;
		case 'h':
		default:
			usage();
			return 1;
		}
	}


	r = ubertooth_connect(ut, ubertooth_device);
	if (r < 0) {
		usage();
		return 1;
	}

	r = ubertooth_check_api(ut);
	if (r < 0)
		return 1;

	/* Clean up on exit. */
	register_cleanup_handler(ut, 1);

	if (do_follow && do_promisc) {
		printf("Error: must choose either -f or -p, one or the other pal\n");
		return 1;
	}

	if (do_follow || do_promisc) {
		usb_pkt_rx rx;

		r = cmd_set_jam_mode(ut->devh, jam_mode);
		if (jam_mode != JAM_NONE && r != 0) {
			printf("Jamming not supported\n");
			return 1;
		}
		cmd_set_modulation(ut->devh, MOD_BT_LOW_ENERGY);

		if (do_follow) {
			u16 channel;
			if (do_adv_index == 37)
				channel = 2402;
			else if (do_adv_index == 38)
				channel = 2426;
			else
				channel = 2480;
			cmd_set_channel(ut->devh, channel);
			cmd_btle_sniffing(ut->devh, 2);
		} else {
			cmd_btle_promisc(ut->devh);
		}

		while (1) {
			int r = cmd_poll(ut->devh, &rx);
			if (r < 0) {
				printf("USB error\n");
				break;
			}
			if (r == sizeof(usb_pkt_rx)) {
				fifo_push(ut->fifo, &rx);
				cb_btle(ut, &cb_opts);
			}
			usleep(500);
		}
		ubertooth_stop(ut);
	}

	if (do_get_aa) {
		access_address = cmd_get_access_address(ut->devh);
		printf("Access address: %08x\n", access_address);
		return 0;
	}

	if (do_set_aa) {
		cmd_set_access_address(ut->devh, access_address);
		printf("access address set to: %08x\n", access_address);
	}

	if (do_crc >= 0) {
		int r;
		if (do_crc == 2) {
			r = cmd_get_crc_verify(ut->devh);
		} else {
			cmd_set_crc_verify(ut->devh, do_crc);
			r = do_crc;
		}
		printf("CRC: %sverify\n", r ? "" : "DO NOT ");
	}

	if (do_slave_mode) {
		u16 channel;
		if (do_adv_index == 37)
			channel = 2402;
		else if (do_adv_index == 38)
			channel = 2426;
		else
			channel = 2480;
		cmd_set_channel(ut->devh, channel);

		cmd_btle_slave(ut->devh, mac_address);
	}

	if (do_target) {
		r = cmd_btle_set_target(ut->devh, mac_address);
		if (r == 0) {
			int i;
			printf("target set to: ");
			for (i = 0; i < 5; ++i)
				printf("%02x:", mac_address[i]);
			printf("%02x\n", mac_address[5]);
		}
	}

	if (!(do_follow || do_promisc || do_get_aa || do_set_aa ||
				do_crc >= 0 || do_slave_mode || do_target))
		usage();

	return 0;
}
